### PR TITLE
enable loading from /etc/ppp/openssl.cnf not only engine settings. 

### PR DESCRIPTION
**Pull request by Sander80**
Loading conf file before first initializing context

-------------------------------

There was another problem we encountered. On Ubuntu 20.04 the default security level is 2. 
https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_security_level.html

But our server has now security level 1.

So it cannot connect.

This can be fixed on the client side with:

```
# /etc/ppp/openssl.cnf - Openssl configuration file
# $Id$

# Manual: config(5)

# Created: ?? ??? 25 10:05:52 2020
#      by: The L2TP IPsec VPN Manager application version 2.beta
#
# WARNING! All changes made in this file will be lost!
#
openssl_conf = openssl_def

[ openssl_def ]
engines = engine_section
ssl_conf = ssl_sect

[ engine_section ]
pkcs11 = pkcs11_section

[ pkcs11_section ]
engine_id = pkcs11
dynamic_path = /usr/lib/x86_64-linux-gnu/engines-1.1/pkcs11.so
MODULE_PATH = /usr/lib/librtpkcs11ecp.so
init=0

[ssl_sect]
system_default = ssl_default_sect

[ssl_default_sect]
CipherString = DEFAULT:@SECLEVEL=1
```

---------------------------------

However it's not nice to patch /etc/sss/openssl.cnf

But before this fix not all options of /etc/ppp/openssl.cnf are loaded

### CODE REVIEW

**Comment by jjkeijser** [comment]
so you're saying that when you move the line
  ctx = SSL_CTX_new(TLS_method());
to *after* the loading of the config+engine then it works with security level set to 2?

The whole concept of OpenSSL's security level (or how they implemented it in CentOS 8/Fedora 30+) still puzzles me ....

**Comment by lo1ol** [comment]
Yes.
But the problem is not only in securiry level. If the patch is not applied, the pppd firstly tries to connect to a server without /etc/ppp/openssl.cnf info (except engine). This happens because  first context initializes only via global openssl.cnf (because the line "ctx = SSL_CTX_new(TLS_method());" is put before reading local openssl.cnf). But if we put initializing context after reading /etc/ppp/openssl.cnf, then the options from the local openssl configuration will be applied to the context.

In our task we are faced with the following problem. We wanted to connect to the server using key with small length (1024 bits), but in the log we got the message that key len was too small. This happened because in Ubuntu default security level is 2 and all rsa keys with length less then 2048 are restricted.

**Comment by jjkeijser** [comment]
sounds like you could add a patch yourself to do a call to "SSL_CTX_set_security_level(1)"  yourself but I don't think that should be part of my patch - esp as the default value for "stock" OpenSSL 1.1.1 is "1" already.    And I don't think it should be added as a configurable option - degrading security is **supposed to be** ugly and dirty.

**Comment by Sander80** [comment]
Right, we surely can make a patch to call SSL_CTX_set_security_level(1) on our side. Or increase security level on the server side (however it's not that easy because requires reissuing new certificates). However please note, that the suggested patch is not decreasing security at all!

Imagine the situation. One is using ppp on Linux, works with Ubuntu 19.10, upgrades to Ubuntu 20.04.. and suddenly his connection stops working. What can he do in this case?
1) Have a patch for ppp that calls SSL_CTX_set_security_level(1) - requires a special version of ppp for him
2) Make a change in /etc/ssl/openssl.cnf - ugly since decreases the security level for all the system
3) Make a change in /etc/ppp/openssl.cnf ????

No, option 3 won't work without our patch. Your code is reading /etc/ppp/openssl.cnf, but doing it not at the right time.

-----------------------------------------------------------------------------------------------------------

Let's even forget about security level. You are trying to read /etc/ppp/openssl.cnf anyway. So you should either forget it at all, or let the user prodive whatever options inside /etc/ppp/openssl.cnf. And without the patch most options in /etc/ppp/openssl.cnf will be ignored. For example, the proposed change to decrease the security level works if the user makes it in /etc/ssl/openssl, but does not work if made in /etc/ppp/openssl..cnf without the patch, but with the patch it works.

-------------------------------------

To summarise - we are not enforcing any options with this suggestions. We are only allowing the user to make changes in /etc/ppp/openssl.cnf that will be used by ppp. I think, the user should have right for this. And anyway if this approach does not work for him, he will go and make changes to /etc/ssl/openssl.cnf, which is worse. What do you think?

**Comment by Neustradamus** [comment]
@jjkeijser, @Sander80, @lo1ol

Thanks guys for your talks, improvements!

Do not forget that there is LibreSSL too.

**Comment by lo1ol** [comment]
Thank you for your response. 

When I made this change, I've decided to not check the existence of local ssl_config, because user may have wanted to load engine described in global openssl config (not in local). So, in this case engine will not be loaded at all.

But if you think that it's better to leave this condition without changes, I can take it back to the old state.

**Comment by jjkeijser** [comment]
I like to play it safe, yes... especially as loading engines (shared libraries) by a program run with root privileges is VERY tricky - all one has to do is patch 'engine_pkcs11.so' to start a terminal and you have root access....

**Comment by Sander80** [comment]
That makes sense. Do you wish me to update the PR with adding "ssl_config &&" ?

**Comment by jjkeijser** [comment]
yes please, the patch looks good otherwise

**Comment by Sander80** [comment]
should be out there already, if I pushed it properly

**Comment by jjkeijser** [comment]
yup I see it, I will merge and test it tomorrow.

**Comment by Sander80** [comment]
Nice )
Now this repository + our tool for vpn connections (https://github.com/Sander80/l2tp-ipsec-vpn) clearly work together without any need for extra patches.

We'll be moitoring your changes, can test them with out tool and also are eager to support the pull request to https://github.com/paulusmack/ppp

**Comment by Neustradamus** [comment]
Good job!

Have you looked about LibreSSL support too?

**Comment by jjkeijser** [comment]
No, and I have no intention of doing so.  The EAP-TLS patch is provided "as is". If it takes a minor patch to compile it with libressl then create a PR.

**Comment by Neustradamus** [comment]
@Sander80: You must add "Issues" part in https://github.com/Sander80/l2tp-ipsec-vpn :)
@jjkeijser: Can you add "Issues" part too for https://github.com/jjkeijser/ppp?

**Comment by Sander80** [comment]
Done

**Comment by Sander80** [comment]
@jjkeijser: sorry for using this place to answer, but I see no other way to write a mesage to you and Neustradamus. I updated my l2tp repository, now it is referring here.

@Neustradamus: seems like gdm85 is not replying and closed the issues. I asked the github team to switch my repository to stand-alone mode from fork. As for libressl, I also have no idea now how to use it and moreover it will have no sense without a support by jjkeikser. So... may be at some point later. For the moment I am solving other things to get l2tp-ipsec-vpn more stable, but soon will have to switch to other duties of mine. However I will try to keep it maintained, at least what relates to bugs, not new features.

**Comment by Neustradamus** [comment]
@Sander80: Thanks for your changes about https://github.com/Sander80/l2tp-ipsec-vpn and about this developement.
-> Issues section is opened :)

I am very very sad about @gdm85 attitude, I do not understand...
